src/qibojit/custom_operators/__init__.py

```diff
@@ -321,25 +333,43 @@ def apply_gate(self, state, gate, nqubits, targets, qubits=None):
         return self.engine.one_qubit_base(state, nqubits, *targets, "apply_gate", qubits, gate)
 
     def apply_x(self, state, nqubits, targets, qubits=None):
-        return self.engine.one_qubit_base(state, nqubits, *targets, "apply_x", qubits)
+        # cast necessary to avoid numba error (missing signature with state complex 64 and gate complex128)
+        gate = self.engine.cast(self.matrices.X, dtype=state.dtype)
```

Comment:
I am not sure if this casting has any significant effect in performane, eg. by creating copies that are not needed, but I think it can be avoided. If am right, the reason we need it here is because qibojit tests fail for numba without it. However, this is an issue with the tests, not the code. Particularly the precision switch is not implemented properly in tests, as we only change the precision of initial state and we do not use `qibo.set_precision()` as we should.

I would suggest trying to fix the tests in qibojit and removing the casting to make sure we are not changing performance for existing backends (numba/cupy). A fix would be to use a fixture for precision in conftest like
```Python
@pytest.fixture
def dtype(precision):
    original_precision = qibo.get_precision()
    qibo.set_precision(precision)
    yield
    qibo.set_precision(original_precision)
```
and check if everything works without this casting. It should work but it also needs some changes in all tests. If you like to give a try let me know, otherwise I can also open another PR for this.

Reply:
I understand, by changing only the precision of the initial state we get the numba error since the matrices are always in double precision. I can give this a try. Thanks for the feedback!

Reply:
Indeed the matrix type is the issue. If everything works properly, doing `qibo.set_precision("single")` should automatically reallocate all the matrices (`matrices.X`, `matrices.Y`, etc.) in single precision so the cast inside each custom operator will not be needed.

Reply:
I was able to implement the pytest fixture, now all tests pass without casting inside the custom operators.
@stavros11 let me you if you like this implementation or if you had something else in mind.